### PR TITLE
[ML] Add api to update trained model deployment number_of_allocations

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This commit adds a new API that users can use calling:

```
POST _ml/trained_models/{model_id}/deployment/_update
{
  "number_of_allocations": 4
}
```

This allows a user to update the number of allocations for a deployment that is `started`.

If the allocations are increased we rebalance and let the assignment planner find how to allocate the additional allocations.

If the allocations are decreased we cannot use the assignment planner. Instead, we implement the reduction in a new class `AllocationReducer` that tries to reduce the allocations so that:

  1. availability zone balance is maintained
  2. assignments that can be completely stopped are preferred to release memory

### CODE REVIEW

File: docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+Requires the `manage_ml` cluster privilege. This privilege is included in the
+`machine_learning_admin` built-in role.
+
+[[ml-update-trained-model-deployment-desc]]
+== {api-description-title}
+
+You can update a trained model deployment whose `assignment_state` is `started`.
+You can either increase or decrease the number of allocations of such a deployment.
+
+[[ml-update-datafeed-path-parms]]
```
Comment (lcawl) [inline docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc:31]:
```suggestion
[[ml-update-trained-model-deployments-path-parms]]
```

File: docs/reference/ml/trained-models/apis/ml-trained-models-apis.asciidoc [hunk 2]
```diff
@@ -15,6 +15,7 @@ You can use the following APIs to perform model management operations:
 * <<infer-trained-model>>
 * <<start-trained-model-deployment>>
 * <<stop-trained-model-deployment>>
+* <<update-trained-model-deployment>>
```
Comment (lcawl) [inline docs/reference/ml/trained-models/apis/ml-trained-models-apis.asciidoc:18]:
```suggestion
* <<ml-update-trained-model-deployment>>
```

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.update_trained_model_deployment.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            "model_id":{
+              "type":"string",
+              "description":"The unique identifier of the trained model."
+            }
+          }
+        }
+      ]
+    },
+    "body":{
+      "description":"The datafeed update settings",
```
Comment (lcawl) [inline rest-api-spec/src/main/resources/rest-api-spec/api/ml.update_trained_model_deployment.json:30]:
```suggestion
      "description":"The updated trained model deployment settings",
```

File: docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+= Update trained model deployment API
+
+[subs="attributes"]
+++++
+<titleabbrev>Update trained model deployment</titleabbrev>
+++++
+
+Updates certain properties of a trained model deployment.
+
+
```
Comment (lcawl) [inline docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc:12]:
```suggestion
beta::[]

```

File: docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+== {api-path-parms-title}
+
+`<model_id>`::
+(Required, string)
+include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=model-id]
+
+[[ml-update-trained-model-deployment-request-body]]
+== {api-request-body-title}
+
+The following property can be updated after the trained model deployment is started:
```
Comment (lcawl) [inline docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc:42]: IMO this text is unnecessary, since it's already mentioned in the description:
```suggestion

```

**Review by lcawl — APPROVED** [review]
One suggestion, otherwise docs LGTM

**Comment by dimitris-athanasiou** [comment]
@lcawl Thank you the fixes! I removed the `ml-` prefix from the labels so that it is consistent with the other docs for trained models. I also removed that redundant sentence.

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/PyTorchModelIT.java [hunk 6]
```diff
@@ -877,14 +877,94 @@ public void testStartDeployment_GivenNoProcessorsLeft_AndLazyStartEnabled() thro
...
+        assertBusy(() -> assertAllocationCount(modelId, 2));
+    }
+
+    public void testUpdateDeployment_GivenAllocationsAreIncreasedOverResources_AndScalingIsPossible() throws Exception {
+        Request loggingSettings = new Request("PUT", "_cluster/settings");
+        loggingSettings.setJsonEntity("""
+            {"persistent" : {
+                    "xpack.ml.max_lazy_ml_nodes": 5
+                }}""");
+        client().performRequest(loggingSettings);
```
Comment (droberts195) [inline x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/PyTorchModelIT.java:906]: This isn't a logging setting, and it doesn't get reset after the test.

I think there's a danger that this will cause some other test to unexpectedly fail depending on the order that tests run.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+---
+"Test with unknown model id":
+  - do:
+      catch: missing
```
Comment (droberts195) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml:4]: Since the only YAML test throws an exception the clients flight recorder won't be able to record the success response format.

Would it be possible to incorporate an update into the middle of one of the tests in `3rd_party_deployment.yml` that starts a deployment?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java [hunk 8]
```diff
@@ -555,6 +556,148 @@ private boolean isScalingPossible(List<DiscoveryNode> nodes) {
...
+    }
+
+    private void decreaseNumberOfAllocations(
+        ClusterState clusterState,
+        TrainedModelAssignment assignment,
+        int numberOfAllocations,
+        ActionListener<ClusterState> listener
+    ) {
+        TrainedModelAssignment.Builder updatedAssignment = numberOfAllocations < assignment.totalTargetAllocations()
+            ? new AllocationReducer(assignment, nodeAvailabilityZoneMapper.getMlNodesByAvailabilityZone()).reduceTo(numberOfAllocations)
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java:688]: There's a potential discrepancy in cluster states used for different parts of the code here.  `nodeAvailabilityZoneMapper` keeps itself up to date by listening for cluster state updates.  But this method runs as part of a cluster state update that is supposed to work off of an exact version of the cluster state.  Maybe `NodeAvailabilityZoneMapper` should have a method that can calculate and return ML nodes by availability zone from a specific `ClusterState` that's passed to it?  Or maybe there's a better way to solve this problem.

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/PyTorchModelIT.java (hunk 6 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/PyTorchModelIT.java:906]: There is a setup method in that test suite that unsets this setting as we also set it in other tests.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml (hunk 7 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml:4]: Definitely. I'll add one.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java (hunk 8 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java:688]: I have raised https://github.com/elastic/elasticsearch/pull/90770 to address this.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AllocationReducer.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            if (largestZoneAssignmentsInAscendingOrder.isEmpty()) {
+                logger.warn("no assignments remain in the largest zone");
+                throw new IllegalStateException("no assignments remain in the largest zone");
+            }
+
+            Map.Entry<String, Integer> smallestAssignmentInLargestZone = largestZoneAssignmentsInAscendingOrder.get(0);
+            if (smallestAssignmentInLargestZone.getValue() == 1
+                || (smallestAssignmentInLargestZone.getValue() <= allocationsToRemove
+                    && (minAllocationsInOtherZones == 0
+                        || ((largestZoneAllocations - smallestAssignmentInLargestZone.getValue()) >= minAllocationsInOtherZones + 1)))) {
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AllocationReducer.java:101]:
```suggestion
            boolean noAllocationsInOtherZones = allocationsPerZoneInAscendingOrder.size() == 1;
            if (smallestAssignmentInLargestZone.getValue() == 1
                || (smallestAssignmentInLargestZone.getValue() <= allocationsToRemove
                    && (noAllocationsInOtherZones
                        || ((largestZoneAllocations - smallestAssignmentInLargestZone.getValue()) > minAllocationsInOtherZones)))) {
```

I found this quite hard to follow could you add a comment please

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AllocationReducer.java (hunk 9 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AllocationReducer.java:101]: I have refactored the check in a method with commentary. That should make it simpler to understand.

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java (hunk 8 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java]: I have now rebased this PR and am using the new method where we pass the specific cluster state

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml (hunk 7 above)
Comment (droberts195) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml:4]: I couldn't see where this was added.

Comment (dimitris-athanasiou) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml:4]: Sorry, I forgot to pick that commit from another branch where I added it. It should be present now in `3rd_party_deployment.yml`.

**Review by droberts195 — APPROVED** [review]
LGTM
